File: Low Level Design (LLD)/Problems/06. Splitwise/Java/Main.java
import java.util.*;

import model.*;
import service.*;

public class Main {
	private static final Scanner scanner = new Scanner(System.in);
	private static final UserService userService = new UserService();
	private static final GroupService groupService = new GroupService();
	private static final ExpenseService expenseService = new ExpenseService();

	public static void main(String[] args) {
		while (true) {
			System.out.println("\n--- Splitwise Menu ---");
			System.out.println("1. Create User");
			System.out.println("2. Create Group");
			System.out.println("3. Add User to Group");
			System.out.println("4. Create Expense");
			System.out.println("5. Show Group Expenses");
			System.out.println("6. Show User Balance Sheet");
			System.out.println("0. Exit");
			System.out.print("Choose: ");
			int choice = scanner.nextInt();
			scanner.nextLine(); // Consume newline

			switch (choice) {
				case 1 -> createUser();
				case 2 -> createGroup();
				case 3 -> addUserToGroup();
				case 4 -> createExpense();
				case 5 -> showGroupExpenses();
				case 6 -> showUserBalanceSheet();
				case 0 -> {
					System.out.println("Goodbye!");
					return;
				}
				default -> System.out.println("Invalid option.");
			}
		}
	}

	private static void createUser() {
		System.out.print("User ID: ");
		String id = scanner.nextLine();
		System.out.print("Name: ");
		String name = scanner.nextLine();
		System.out.print("Email: ");
		String email = scanner.nextLine();
		System.out.print("Phone: ");
		String phone = scanner.nextLine();
		userService.createUser(id, name, email, phone);
		System.out.println("User created.");
	}

	private static void createGroup() {
		System.out.print("Group ID: ");
		String groupId = scanner.nextLine();
		System.out.print("Group Name: ");
		String name = scanner.nextLine();
		groupService.createGroup(groupId, name);
		System.out.println("Group created.");
	}

	private static void addUserToGroup() {
		System.out.print("Group ID: ");
		String groupId = scanner.nextLine();
		System.out.print("User ID: ");
		String userId = scanner.nextLine();
		User user = userService.getUser(userId);
		if (user != null) {
			groupService.addUserToGroup(groupId, user);
			System.out.println("User added to group.");
		} else {
			System.out.println("User not found.");
		}
	}

	private static void createExpense() {
		System.out.print("Group ID: ");
		String groupId = scanner.nextLine();
		System.out.print("Description: ");
		String desc = scanner.nextLine();
		System.out.print("Amount: ");
		double amount = scanner.nextDouble();
		scanner.nextLine();
		System.out.print("Paid by (userId): ");
		String paidById = scanner.nextLine();

		System.out.print("Enter expense type (EQUAL, EXACT, PERCENTAGE): ");
		String type = scanner.nextLine().toUpperCase();

		Group group = groupService.createGroup(groupId, groupId);
		List<User> participants = new ArrayList<>();
		for (User user : group.getMembers()) {
			System.out.print("Include " + user.getName() + "? (y/n): ");
			if (scanner.nextLine().equalsIgnoreCase("y")) {
				participants.add(user);
			}
		}

		List<Double> amounts = null;
		if (type.equals("EXACT")) {
			amounts = new ArrayList<>();
			for (User user : participants) {
				System.out.print("Amount for " + user.getName() + ": ");
				amounts.add(scanner.nextDouble());
				scanner.nextLine();
			}
		} else if (type.equals("PERCENTAGE")) {
			amounts = new ArrayList<>();
			for (User user : participants) {
				System.out.print("Percentage for " + user.getName() + ": ");
				amounts.add(scanner.nextDouble());
				scanner.nextLine();
			}
		}

		User paidBy = userService.getUser(paidById);
		Expense expense = null;
		String expenseId = UUID.randomUUID().toString();

		switch (type) {
			case "EQUAL" ->
					expense = new EqualExpense(expenseId, desc, amount, paidBy, participants);
			case "EXACT" ->
					expense = new ExactExpense(expenseId, desc, amount, paidBy, participants, amounts);
			case "PERCENTAGE" ->
					expense = new PercentageExpense(expenseId, desc, amount, paidBy, participants, amounts);
			default -> System.out.println("Invalid expense type.");
		}

		if (expense != null) {
			expenseService.createExpense(expense);
			groupService.createGroup(groupId, "").addExpense(expense);
			System.out.println("Expense recorded.");
		}
	}

	private static void showGroupExpenses() {
		System.out.print("Group ID: ");
		String groupId = scanner.nextLine();
		List<Expense> expenses = groupService.getGroupExpenses(groupId);
		for (Expense e : expenses) {
			System.out.println("\n[" + e.getExpenseType() + "] " + e.getDescription() + " - $" + e.getAmount());
			for (Map.Entry<User, Double> entry : e.getUserShare().entrySet()) {
				System.out.println("  " + entry.getKey().getName() + " owes: $" + entry.getValue());
			}
		}
	}

	private static void showUserBalanceSheet() {
		System.out.print("User ID: ");
		String userId = scanner.nextLine();
		User user = userService.getUser(userId);
		if (user != null) {
			System.out.println("Balance Sheet for " + user.getName() + ":");
			for (Map.Entry<String, Double> entry : user.getBalanceSheet().entrySet()) {
				System.out.printf("  Owes to %s: %.2f%n", entry.getKey(), entry.getValue());
			}
		} else {
			System.out.println("User not found.");
		}
	}
}

File: Low Level Design (LLD)/Problems/06. Splitwise/Cpp/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>
#include <unordered_map>
#include <limits>
#include <random>
#include <chrono>

#include "./service/include/UserService.h"
#include "./service/include/GroupService.h"
#include "./service/include/ExpenseService.h"
#include "./model/include/EqualExpense.h"
#include "./model/include/ExactExpense.h"
#include "./model/include/PercentageExpense.h"

using namespace std;

UserService userService;
GroupService groupService;
ExpenseService expenseService;

string generateUUID() {
    static random_device rd;
    static mt19937 gen(rd());
    static uniform_int_distribution<> dis(0, 15);
    static const char* hex = "0123456789abcdef";

    string uuid = "";
    for (int i = 0; i < 32; ++i) {
        uuid += hex[dis(gen)];
        if (i == 7 || i == 11 || i == 15 || i == 19)
            uuid += '-';
    }
    return uuid;
}

void createUser() {
    string id, name, email, phone;
    cout << "User ID: ";
    getline(cin, id);
    cout << "Name: ";
    getline(cin, name);
    cout << "Email: ";
    getline(cin, email);
    cout << "Phone: ";
    getline(cin, phone);

    userService.createUser(id, name, email, phone);
    cout << "User created." << endl;
}

void createGroup() {
    string groupId, name;
    cout << "Group ID: ";
    getline(cin, groupId);
    cout << "Group Name: ";
    getline(cin, name);

    groupService.createGroup(groupId, name);
    cout << "Group created." << endl;
}

void addUserToGroup() {
    string groupId, userId;
    cout << "Group ID: ";
    getline(cin, groupId);
    cout << "User ID: ";
    getline(cin, userId);

    auto user = userService.getUser(userId);
    if (user) {
        groupService.addUserToGroup(groupId, user);
        cout << "User added to group." << endl;
    } else {
        cout << "User not found." << endl;
    }
}

void createExpense() {
    string groupId, desc, paidById, type;
    double amount;

    cout << "Group ID: ";
    getline(cin, groupId);

    auto group = groupService.getGroup(groupId);
    if (!group) {
        cout << "Group not found." << endl;
        return;
    }

    cout << "Description: ";
    getline(cin, desc);
    cout << "Amount: ";
    cin >> amount;
    cin.ignore();

    cout << "Paid by (userId): ";
    getline(cin, paidById);

    cout << "Enter expense type (EQUAL, EXACT, PERCENTAGE): ";
    getline(cin, type);

    vector<shared_ptr<User>> participants;
    for (const auto& user : group->getMembers()) {
        cout << "Include " << user->getName() << "? (y/n): ";
        string input;
        getline(cin, input);
        if (input == "y" || input == "Y") {
            participants.push_back(user);
        }
    }

    vector<double> amounts;
    if (type == "EXACT" || type == "PERCENTAGE") {
        for (const auto& user : participants) {
            double val;
            cout << (type == "EXACT" ? "Amount" : "Percentage") << " for " << user->getName() << ": ";
            cin >> val;
            cin.ignore();
            amounts.push_back(val);
        }
    }

    auto paidBy = userService.getUser(paidById);
    if (!paidBy) {
        cout << "Invalid user ID." << endl;
        return;
    }

    string expenseId = generateUUID();
    shared_ptr<Expense> expense;

    if (type == "EQUAL") {
        expense = make_shared<EqualExpense>(expenseId, desc, amount, paidBy, participants);
    } else if (type == "EXACT") {
        expense = make_shared<ExactExpense>(expenseId, desc, amount, paidBy, participants, amounts);
    } else if (type == "PERCENTAGE") {
        expense = make_shared<PercentageExpense>(expenseId, desc, amount, paidBy, participants, amounts);
    } else {
        cout << "Invalid expense type." << endl;
        return;
    }

    expenseService.createExpense(expense);
    group->addExpense(expense);

    cout << "Expense recorded." << endl;
}

void showGroupExpenses() {
    string groupId;
    cout << "Group ID: ";
    getline(cin, groupId);

    auto expenses = groupService.getGroupExpenses(groupId);
    for (const auto& e : expenses) {
        cout << "\n[" << static_cast<int>(e->getExpenseType()) << "] " << e->getDescription()
             << " - $" << e->getAmount() << endl;
        for (const auto& entry : e->getUserShare()) {
            cout << "  " << entry.first->getName() << " owes: $" << entry.second << endl;
        }
    }
}

void showUserBalanceSheet() {
    string userId;
    cout << "User ID: ";
    getline(cin, userId);

    auto user = userService.getUser(userId);
    if (!user) {
        cout << "User not found." << endl;
        return;
    }

    cout << "Balance Sheet for " << user->getName() << ":\n";
    for (const auto& entry : user->getBalanceSheet()) {
        cout << "  Owes to " << entry.first << ": " << entry.second << endl;
    }
}

int main() {
    while (true) {
        cout << "\n--- Splitwise Menu ---\n";
        cout << "1. Create User\n";
        cout << "2. Create Group\n";
        cout << "3. Add User to Group\n";
        cout << "4. Create Expense\n";
        cout << "5. Show Group Expenses\n";
        cout << "6. Show User Balance Sheet\n";
        cout << "0. Exit\n";
        cout << "Choose: ";

        int choice;
        cin >> choice;
        cin.ignore(numeric_limits<streamsize>::max(), '\n'); // Clear input buffer

        switch (choice) {
            case 1: createUser(); break;
            case 2: createGroup(); break;
            case 3: addUserToGroup(); break;
            case 4: createExpense(); break;
            case 5: showGroupExpenses(); break;
            case 6: showUserBalanceSheet(); break;
            case 0:
                cout << "Goodbye!" << endl;
                return 0;
            default:
                cout << "Invalid option." << endl;
        }
    }
}
